### PR TITLE
Preserve thread context when calling send test message

### PR DESCRIPTION
**Pull request by joshuali925**
### Description
Restore thread context inside coroutine so that messages are sent with user tenant, and event for test messages will go to the correct tenant.

### Issues Resolved
partially #244 

### Check List
- [ ] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by zhongnansu** [comment]
1. Is this issue specific with send test message? 
2. Does your change also fix non-test notifications?

**Comment by joshuali925** [comment]
> 1. Is this issue specific with send test message?
> 2. Does your change also fix non-test notifications?

No. The issue exists for test message, send notification in transport action, and send notification in job-scheduler. They need three different solutions. This PR only fixes for test message.

---
@akbhatta I just realized after adding another layer for test message, send test message is similar to other actions. Do we still need to restore context? I tested without it and tenant still seems right.
https://github.com/opensearch-project/notifications/blob/7e327db78b17ac0ec4d889f329ace2b8fe9d9e03/notifications/notifications/src/main/kotlin/org/opensearch/notifications/action/PluginBaseAction.kt#L79-L85

File: notifications/notifications/src/main/kotlin/org/opensearch/notifications/action/SendTestNotificationAction.kt [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            source,
+            message,
+            channelIds,
+            object : ActionListener<SendNotificationResponse> {
+                override fun onResponse(p0: SendNotificationResponse) {
+                    log.info("$LOG_PREFIX:SendTestNotificationAction-send:${p0.notificationId}")
+                    listener.onResponse(p0)
+                }
+
+                override fun onFailure(p0: java.lang.Exception) {
```
Comment (akbhatta) [inline notifications/notifications/src/main/kotlin/org/opensearch/notifications/action/SendTestNotificationAction.kt:62]: nit: add import and rename variable name?

**Review by akbhatta — APPROVED** [review]

**Comment by akbhatta** [comment]
> > 1. Is this issue specific with send test message?
> > 2. Does your change also fix non-test notifications?
> 
> No. The issue exists for test message, send notification in transport action, and send notification in job-scheduler. They need three different solutions. This PR only fixes for test message.
> 
> @akbhatta I just realized after adding another layer for test message, send test message is similar to other actions. Do we still need to restore context? I tested without it and tenant still seems right.
> 
> https://github.com/opensearch-project/notifications/blob/7e327db78b17ac0ec4d889f329ace2b8fe9d9e03/notifications/notifications/src/main/kotlin/org/opensearch/notifications/action/PluginBaseAction.kt#L79-L85

We don't require but keeping it is better as we are keeping security context.

**Review by akbhatta — APPROVED** [review]

**Pull request merged**